File: archive/v2-python/kite/utils/metrics.py
"""
Metrics collection for Kite library.

Provides Prometheus-compatible metrics for monitoring scraper performance
and health.
"""

from typing import Dict, Optional, Any
from prometheus_client import Counter, Histogram, Gauge, Info
import time
from functools import wraps
from contextlib import contextmanager


# ===== Request Metrics =====
request_counter = Counter(
    "kite_requests_total",
    "Total number of HTTP requests made",
    ["scraper", "method", "status"],
)

request_duration = Histogram(
    "kite_request_duration_seconds",
    "HTTP request duration in seconds",
    ["scraper", "method"],
    buckets=(0.1, 0.5, 1.0, 2.0, 5.0, 10.0, 30.0, 60.0),
)

request_errors = Counter(
    "kite_request_errors_total",
    "Total number of request errors",
    ["scraper", "error_type"],
)

# ===== Scraper Metrics =====
cases_scraped = Counter(
    "kite_cases_scraped_total",
    "Total number of cases successfully scraped",
    ["scraper", "jurisdiction"],
)

scraping_duration = Histogram(
    "kite_scraping_duration_seconds",
    "Time taken to scrape cases",
    ["scraper"],
    buckets=(1.0, 5.0, 10.0, 30.0, 60.0, 120.0, 300.0),
)

active_scrapers = Gauge(
    "kite_active_scrapers",
    "Number of currently active scrapers",
    ["scraper"],
)

# ===== Rate Limiting Metrics =====
rate_limit_hits = Counter(
    "kite_rate_limit_hits_total",
    "Number of times rate limiting was triggered",
    ["scraper"],
)

rate_limit_wait_time = Histogram(
    "kite_rate_limit_wait_seconds",
    "Time spent waiting due to rate limiting",
    ["scraper"],
    buckets=(0.1, 0.5, 1.0, 2.0, 5.0, 10.0),
)

# ===== Parsing Metrics =====
parsing_errors = Counter(
    "kite_parsing_errors_total",
    "Total number of parsing errors",
    ["scraper", "error_type"],
)

parsing_duration = Histogram(
    "kite_parsing_duration_seconds",
    "Time taken to parse case data",
    ["scraper"],
    buckets=(0.01, 0.05, 0.1, 0.5, 1.0, 5.0),
)

# ===== Application Info =====
app_info = Info("kite_app", "Application information")
app_info.info(
    {
        "version": "1.0.0",
        "name": "Kite",
        "description": "Legal case law scraper library",
    }
)


class MetricsCollector:
    """
    Metrics collector for tracking scraper operations.

    Provides context managers and decorators for automatic metrics collection.
    """

    def __init__(self, scraper_name: str, jurisdiction: Optional[str] = None):
        """
        Initialize metrics collector.

        Args:
            scraper_name: Name of the scraper
            jurisdiction: Jurisdiction being scraped
        """
        self.scraper_name = scraper_name
        self.jurisdiction = jurisdiction or "unknown"

    @contextmanager
    def track_request(
        self, method: str = "GET", url: Optional[str] = None
    ) -> Any:
        """
        Context manager to track HTTP request metrics.

        Args:
            method: HTTP method
            url: Request URL (optional)

        Yields:
            Dict with request metadata

        Example:
            >>> collector = MetricsCollector("CourtListener")
            >>> with collector.track_request("GET") as req:
            ...     response = make_request()
            ...     req["status"] = response.status_code
        """
        start_time = time.time()
        request_data = {"status": "unknown", "error": None}

        try:
            yield request_data
            status = request_data.get("status", "unknown")
            request_counter.labels(
                scraper=self.scraper_name, method=method, status=status
            ).inc()

        except Exception as e:
            request_data["error"] = type(e).__name__
            request_errors.labels(
                scraper=self.scraper_name, error_type=type(e).__name__
            ).inc()
            raise

        finally:
            duration = time.time() - start_time
            request_duration.labels(scraper=self.scraper_name, method=method).observe(
                duration
            )

    @contextmanager
    def track_scraping(self) -> Any:
        """
        Context manager to track scraping operation metrics.

        Yields:
            Dict with scraping metadata

        Example:
            >>> collector = MetricsCollector("CourtListener", "US")
            >>> with collector.track_scraping() as scrape:
            ...     cases = scrape_cases()
            ...     scrape["cases_count"] = len(cases)
        """
        start_time = time.time()
        scrape_data = {"cases_count": 0}

        # Increment active scrapers
        active_scrapers.labels(scraper=self.scraper_name).inc()

        try:
            yield scrape_data

            # Record successful scraping
            count = scrape_data.get("cases_count", 0)
            if count > 0:
                cases_scraped.labels(
                    scraper=self.scraper_name, jurisdiction=self.jurisdiction
                ).inc(count)

        finally:
            # Decrement active scrapers
            active_scrapers.labels(scraper=self.scraper_name).dec()

            # Record duration
            duration = time.time() - start_time
            scraping_duration.labels(scraper=self.scraper_name).observe(duration)

    @contextmanager
    def track_parsing(self) -> Any:
        """
        Context manager to track parsing operation metrics.

        Yields:
            Dict with parsing metadata

        Example:
            >>> collector = MetricsCollector("CourtListener")
            >>> with collector.track_parsing() as parse:
            ...     case_data = parse_html(html)
        """
        start_time = time.time()
        parse_data = {"error": None}

        try:
            yield parse_data

        except Exception as e:
            parse_data["error"] = type(e).__name__
            parsing_errors.labels(
                scraper=self.scraper_name, error_type=type(e).__name__
            ).inc()
            raise

        finally:
            duration = time.time() - start_time
            parsing_duration.labels(scraper=self.scraper_name).observe(duration)

    def record_rate_limit(self, wait_time: float) -> None:
        """
        Record rate limiting event.

        Args:
            wait_time: Time spent waiting (seconds)
        """
        rate_limit_hits.labels(scraper=self.scraper_name).inc()
        rate_limit_wait_time.labels(scraper=self.scraper_name).observe(wait_time)


def track_function_duration(scraper_name: str, operation: str = "operation"):
    """
    Decorator to track function execution duration.

    Args:
        scraper_name: Name of the scraper
        operation: Name of the operation being tracked

    Example:
        >>> @track_function_duration("CourtListener", "search")
        ... def search_cases(query):
        ...     return perform_search(query)
    """

    def decorator(func):
        @wraps(func)
        def wrapper(*args, **kwargs):
            start_time = time.time()
            try:
                result = func(*args, **kwargs)
                return result
            finally:
                duration = time.time() - start_time
                scraping_duration.labels(scraper=scraper_name).observe(duration)

        return wrapper

    return decorator


# Utility functions for common metrics operations
def increment_request_counter(
    scraper_name: str, method: str = "GET", status: int = 200
) -> None:
    """
    Increment request counter.

    Args:
        scraper_name: Name of the scraper
        method: HTTP method
        status: HTTP status code
    """
    request_counter.labels(
        scraper=scraper_name, method=method, status=str(status)
    ).inc()


def record_request_duration(
    scraper_name: str, method: str, duration: float
) -> None:
    """
    Record request duration.

    Args:
        scraper_name: Name of the scraper
        method: HTTP method
        duration: Duration in seconds
    """
    request_duration.labels(scraper=scraper_name, method=method).observe(duration)


def increment_error_counter(scraper_name: str, error_type: str) -> None:
    """
    Increment error counter.

    Args:
        scraper_name: Name of the scraper
        error_type: Type of error
    """
    request_errors.labels(scraper=scraper_name, error_type=error_type).inc()

File: archive/v3-nim/src/kite/metrics.nim
import std/[strutils, tables, sequtils, times, asyncdispatch, asyncnet]
import chronicles
from ./config import getConfig

# Minimal in-memory Prometheus registry with labeled counters, gauges, and histograms

type
  LabelKey = string # canonical key like "scraper=CL;method=GET;status=200"

  Counter = Table[LabelKey, float]
  Gauge = Table[LabelKey, float]

  HistogramDef = object
    buckets: seq[float]
    counts: Table[LabelKey, seq[float]] # per bucket incl +Inf
    sums: Table[LabelKey, float]
    totals: Table[LabelKey, float]

var counters = initTable[string, Counter]()
var gauges = initTable[string, Gauge]()
var histos = initTable[string, HistogramDef]()

proc canon(labels: openArray[(string, string)]): LabelKey =
  var parts = labels.toSeq
  parts.sort(proc (a,b: (string,string)): int = cmp(a[0], b[0]))
  for (k, v) in parts: result.add k & "=" & v & ";"

proc getCounter(name: string): var Counter =
  if not counters.hasKey(name): counters[name] = initTable[LabelKey, float]()
  return counters[name]

proc getGauge(name: string): var Gauge =
  if not gauges.hasKey(name): gauges[name] = initTable[LabelKey, float]()
  return gauges[name]

proc getHisto(name: string; buckets: seq[float]): var HistogramDef =
  if not histos.hasKey(name):
    var h: HistogramDef
    h.buckets = buckets
    h.counts = initTable[LabelKey, seq[float]]()
    h.sums = initTable[LabelKey, float]()
    h.totals = initTable[LabelKey, float]()
    histos[name] = h
  return histos[name]

# === Metric APIs (matching v1 names) ===

# Counters
proc incRequest*(scraper, method, status: string; delta = 1.0) =
  let key = canon({("scraper", scraper), ("method", method), ("status", status)})
  var c = getCounter("kite_requests_total")
  c[key] = c.getOrDefault(key, 0.0) + delta

proc incRequestError*(scraper, errorType: string; delta = 1.0) =
  let key = canon({("scraper", scraper), ("error_type", errorType)})
  var c = getCounter("kite_request_errors_total")
  c[key] = c.getOrDefault(key, 0.0) + delta

proc incCasesScraped*(scraper, jurisdiction: string; delta = 1.0) =
  let key = canon({("scraper", scraper), ("jurisdiction", jurisdiction)})
  var c = getCounter("kite_cases_scraped_total")
  c[key] = c.getOrDefault(key, 0.0) + delta

proc incRateLimitHits*(scraper: string; delta = 1.0) =
  let key = canon({("scraper", scraper)})
  var c = getCounter("kite_rate_limit_hits_total")
  c[key] = c.getOrDefault(key, 0.0) + delta

proc incParsingErrors*(scraper, errorType: string; delta = 1.0) =
  let key = canon({("scraper", scraper), ("error_type", errorType)})
  var c = getCounter("kite_parsing_errors_total")
  c[key] = c.getOrDefault(key, 0.0) + delta

# Gauges
proc gaugeActiveScrapersAdd*(scraper: string; delta = 1.0) =
  let key = canon({("scraper", scraper)})
  var g = getGauge("kite_active_scrapers")
  g[key] = g.getOrDefault(key, 0.0) + delta

# Histograms
proc observe*(name: string; labels: openArray[(string,string)]; value: float; buckets: seq[float]) =
  var h = getHisto(name, buckets)
  let key = canon(labels)
  if not h.counts.hasKey(key):
    h.counts[key] = newSeq[float](buckets.len + 1) # +Inf bucket
    h.sums[key] = 0.0
    h.totals[key] = 0.0
  # increment appropriate bucket
  var idx = buckets.len # default +Inf
  for i, b in buckets:
    if value <= b: idx = i; break
  h.counts[key][idx] = h.counts[key][idx] + 1.0
  h.sums[key] = h.sums[key] + value
  h.totals[key] = h.totals[key] + 1.0

proc observeRequestDuration*(scraper, method: string; seconds: float) =
  observe("kite_request_duration_seconds", {("scraper", scraper), ("method", method)}, seconds, @[0.1,0.5,1.0,2.0,5.0,10.0,30.0,60.0])

proc observeScrapingDuration*(scraper: string; seconds: float) =
  observe("kite_scraping_duration_seconds", {("scraper", scraper)}, seconds, @[1.0,5.0,10.0,30.0,60.0,120.0,300.0])

proc observeRateLimitWait*(scraper: string; seconds: float) =
  observe("kite_rate_limit_wait_seconds", {("scraper", scraper)}, seconds, @[0.1,0.5,1.0,2.0,5.0,10.0])

proc observeParsingDuration*(scraper: string; seconds: float) =
  observe("kite_parsing_duration_seconds", {("scraper", scraper)}, seconds, @[0.01,0.05,0.1,0.5,1.0,5.0])

# App info
proc appInfo*(): string =
  "kite_app_info{version=\"2.0.0\",name=\"Kite\",description=\"Legal case law scraper library\"} 1\n"

# Exposition
proc metricsText*(): string =
  var s = "# HELP kite_app Application information\n# TYPE kite_app info\n"
  s.add appInfo()

  let counterHelp: Table[string,string] = {
    "kite_requests_total": "Total number of HTTP requests made",
    "kite_request_errors_total": "Total number of request errors",
    "kite_cases_scraped_total": "Total number of cases successfully scraped",
    "kite_rate_limit_hits_total": "Number of times rate limiting was triggered",
    "kite_parsing_errors_total": "Total number of parsing errors",
  }.toTable

  for name, store in counters:
    s.add "# HELP " & name & " " & counterHelp.getOrDefault(name, "") & "\n# TYPE " & name & " counter\n"
    for k, v in store:
      # convert LabelKey canon to {k="v",...}
      var labelsOut: seq[string] = @[]
      for part in k.split(';'):
        if part.len == 0: continue
        let kv = part.split('=')
        if kv.len == 2:
          labelsOut.add kv[0] & "=\"" & kv[1] & "\""
      let labelStr = if labelsOut.len > 0: "{" & labelsOut.join(",") & "}" else: ""
      s.add name & labelStr & " " & $v & "\n"

  # Gauges
  for name, store in gauges:
    s.add "# HELP " & name & " " & (if name == "kite_active_scrapers": "Number of currently active scrapers" else: "") & "\n# TYPE " & name & " gauge\n"
    for k, v in store:
      var labelsOut: seq[string] = @[]
      for part in k.split(';'):
        if part.len == 0: continue
        let kv = part.split('=')
        if kv.len == 2:
          labelsOut.add kv[0] & "=\"" & kv[1] & "\""
      let labelStr = if labelsOut.len > 0: "{" & labelsOut.join(",") & "}" else: ""
      s.add name & labelStr & " " & $v & "\n"

  # Histograms
  let histHelp: Table[string,string] = {
    "kite_request_duration_seconds": "HTTP request duration in seconds",
    "kite_scraping_duration_seconds": "Time taken to scrape cases",
    "kite_rate_limit_wait_seconds": "Time spent waiting due to rate limiting",
    "kite_parsing_duration_seconds": "Time taken to parse case data",
  }.toTable

  for name, h in histos:
    s.add "# HELP " & name & " " & histHelp.getOrDefault(name, "") & "\n# TYPE " & name & " histogram\n"
    for k, counts in h.counts:
      # base labels without le
      var baseLabels: seq[(string,string)] = @[]
      for part in k.split(';'):
        if part.len == 0: continue
        let kv = part.split('=')
        if kv.len == 2: baseLabels.add (kv[0], kv[1])
      # cumulative
      var cum = 0.0
      for i in 0 ..< h.buckets.len:
        cum += counts[i]
        var labelsOut = baseLabels
        labelsOut.add ("le", $h.buckets[i])
        let keyStr = "{" & labelsOut.mapIt(it[0] & "=\"" & it[1] & "\"").join(",") & "}"
        s.add name & "_bucket" & keyStr & " " & $cum & "\n"
      # +Inf
      cum += counts[^1]
      var labelsOut2 = baseLabels
      labelsOut2.add ("le", "+Inf")
      let keyStr2 = "{" & labelsOut2.mapIt(it[0] & "=\"" & it[1] & "\"").join(",") & "}"
      s.add name & "_bucket" & keyStr2 & " " & $cum & "\n"
      # _sum and _count
      let sumv = h.sums.getOrDefault(k, 0.0)
      let cntv = h.totals.getOrDefault(k, 0.0)
      let keyBase = if baseLabels.len > 0: "{" & baseLabels.mapIt(it[0] & "=\"" & it[1] & "\"").join(",") & "}" else: ""
      s.add name & "_sum" & keyBase & " " & $sumv & "\n"
      s.add name & "_count" & keyBase & " " & $cntv & "\n"

  return s

# Simple HTTP server exposing /metrics and /health
proc handleClient(sock: AsyncSocket; metricsPath: string) {.async.} =
  let req = await sock.recvLine()
  if req.startsWith("GET "):
    if req.contains(" " & metricsPath & " "):
      let body = metricsText()
      let resp = "HTTP/1.1 200 OK\r\nContent-Type: text/plain; version=0.0.4\r\nContent-Length: " & $body.len & "\r\n\r\n" & body
      await sock.send(resp)
    elif req.contains(" /health "):
      let body = "{\"status\": \"healthy\", \"service\": \"kite\"}"
      let resp = "HTTP/1.1 200 OK\r\nContent-Type: application/json\r\nContent-Length: " & $body.len & "\r\n\r\n" & body
      await sock.send(resp)
    else:
      await sock.send("HTTP/1.1 404 Not Found\r\nContent-Length: 0\r\n\r\n")
  await sock.close()

proc runMetricsServer*() {.async.} =
  let cfg = getConfig()
  let port = cfg.metricsPort
  let path = cfg.metricsPath
  let server = newAsyncSocket()
  server.setSockOpt(OptReuseAddr, true)
  await server.bindAddr(Port(port))
  server.listen()
  info "metrics_server_started", port=port, metrics_path=path, health_path="/health"
  while true:
    let client = await server.accept()
    asyncCheck handleClient(client, path)
